File: sdk/js/webview-bridge.js
/**
 * WebView Bridge JavaScript/TypeScript SDK
 * A simple client for WebView Bridge server.
 * 
 * @example
 * ```javascript
 * import { WebViewBridge, MCPBridge, WebDriverBridge } from './webview-bridge.js';
 * 
 * // Native API
 * const bridge = new WebViewBridge();
 * const sessionId = await bridge.createSession();
 * await bridge.navigate(sessionId, 'https://example.com');
 * const title = await bridge.execute(sessionId, 'return document.title');
 * await bridge.close(sessionId);
 * 
 * // MCP API (for AI agents)
 * const mcp = new MCPBridge();
 * await mcp.browse('https://example.com');
 * const html = await mcp.getHtml();
 * 
 * // WebDriver API (Selenium compatible)
 * const wd = new WebDriverBridge();
 * const session = await wd.newSession();
 * await wd.navigate(session, 'https://example.com');
 * await wd.quit(session);
 * ```
 */

/**
 * WebView Bridge Native API Client
 */
export class WebViewBridge {
    /**
     * @param {string} host - Server host
     * @param {number} port - Server port
     */
    constructor(host = 'localhost', port = 9400) {
        this.baseUrl = `http://${host}:${port}`;
    }

    /**
     * Check server health
     * @returns {Promise<boolean>}
     */
    async health() {
        try {
            const response = await fetch(`${this.baseUrl}/health`);
            return response.ok;
        } catch {
            return false;
        }
    }

    /**
     * Create a new browser session
     * @param {string} profile - Profile name
     * @param {boolean} headless - Headless mode
     * @returns {Promise<string>} Session ID
     */
    async createSession(profile = 'default', headless = false) {
        const response = await fetch(`${this.baseUrl}/create`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ profile, headless })
        });
        if (!response.ok) throw new Error(`Failed to create session: ${response.status}`);
        const data = await response.json();
        return data.id;
    }

    /**
     * Navigate to URL
     * @param {string} sessionId - Session ID
     * @param {string} url - URL to navigate to
     * @returns {Promise<boolean>}
     */
    async navigate(sessionId, url) {
        const response = await fetch(`${this.baseUrl}/navigate/${sessionId}`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ url })
        });
        if (!response.ok) throw new Error(`Navigate failed: ${response.status}`);
        return true;
    }

    /**
     * Execute JavaScript
     * @param {string} sessionId - Session ID
     * @param {string} script - JavaScript code
     * @returns {Promise<any>}
     */
    async execute(sessionId, script) {
        const response = await fetch(`${this.baseUrl}/execute/${sessionId}`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ script })
        });
        if (!response.ok) throw new Error(`Execute failed: ${response.status}`);
        const data = await response.json();
        return data.result;
    }

    /**
     * Get session status
     * @param {string} sessionId - Session ID
     * @returns {Promise<object>}
     */
    async getStatus(sessionId) {
        const response = await fetch(`${this.baseUrl}/status/${sessionId}`);
        if (!response.ok) throw new Error(`Get status failed: ${response.status}`);
        return await response.json();
    }

    /**
     * Get page snapshot
     * @param {string} sessionId - Session ID
     * @param {string} format - 'html' or 'text'
     * @returns {Promise<string>}
     */
    async snapshot(sessionId, format = 'html') {
        const response = await fetch(`${this.baseUrl}/snapshot/${sessionId}?format=${format}`);
        if (!response.ok) throw new Error(`Snapshot failed: ${response.status}`);
        const data = await response.json();
        return data.content;
    }

    /**
     * Take screenshot
     * @param {string} sessionId - Session ID
     * @returns {Promise<string>} Base64 PNG
     */
    async screenshot(sessionId) {
        const response = await fetch(`${this.baseUrl}/screenshot/${sessionId}`);
        if (!response.ok) throw new Error(`Screenshot failed: ${response.status}`);
        const data = await response.json();
        return data.image;
    }

    /**
     * Get cookies
     * @param {string} sessionId - Session ID
     * @returns {Promise<Array>}
     */
    async getCookies(sessionId) {
        const response = await fetch(`${this.baseUrl}/cookies/${sessionId}`);
        if (!response.ok) throw new Error(`Get cookies failed: ${response.status}`);
        const data = await response.json();
        return data.cookies || [];
    }

    /**
     * Wait for selector
     * @param {string} sessionId - Session ID
     * @param {string} selector - CSS selector
     * @param {number} timeoutMs - Timeout in milliseconds
     * @returns {Promise<boolean>}
     */
    async waitForSelector(sessionId, selector, timeoutMs = 10000) {
        const response = await fetch(`${this.baseUrl}/wait/${sessionId}`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ selector, timeout_ms: timeoutMs })
        });
        if (!response.ok) throw new Error(`Wait failed: ${response.status}`);
        const data = await response.json();
        return data.found;
    }

    /**
     * Extract content from elements
     * @param {string} sessionId - Session ID
     * @param {string} selector - CSS selector
     * @param {string} attribute - Attribute to extract
     * @returns {Promise<Array>}
     */
    async extract(sessionId, selector, attribute = 'text') {
        const response = await fetch(`${this.baseUrl}/extract/${sessionId}`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ selector, attribute, all: true })
        });
        if (!response.ok) throw new Error(`Extract failed: ${response.status}`);
        const data = await response.json();
        return data.results || [];
    }

    /**
     * Close session
     * @param {string} sessionId - Session ID
     * @returns {Promise<boolean>}
     */
    async close(sessionId) {
        const response = await fetch(`${this.baseUrl}/close/${sessionId}`, {
            method: 'DELETE'
        });
        if (!response.ok) throw new Error(`Close failed: ${response.status}`);
        return true;
    }
}

/**
 * WebDriver Protocol Compatible Client (Selenium compatible)
 */
export class WebDriverBridge {
    constructor(host = 'localhost', port = 9400) {
        this.baseUrl = `http://${host}:${port}/wd/hub`;
    }

    async newSession() {
        const response = await fetch(`${this.baseUrl}/session`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ capabilities: {} })
        });
        if (!response.ok) throw new Error(`New session failed: ${response.status}`);
        const data = await response.json();
        return data.value.sessionId;
    }

    async navigate(sessionId, url) {
        const response = await fetch(`${this.baseUrl}/session/${sessionId}/url`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ url })
        });
        if (!response.ok) throw new Error(`Navigate failed: ${response.status}`);
    }

    async getUrl(sessionId) {
        const response = await fetch(`${this.baseUrl}/session/${sessionId}/url`);
        if (!response.ok) throw new Error(`Get URL failed: ${response.status}`);
        const data = await response.json();
        return data.value;
    }

    async getTitle(sessionId) {
        const response = await fetch(`${this.baseUrl}/session/${sessionId}/title`);
        if (!response.ok) throw new Error(`Get title failed: ${response.status}`);
        const data = await response.json();
        return data.value;
    }

    async findElement(sessionId, using, value) {
        const response = await fetch(`${this.baseUrl}/session/${sessionId}/element`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ using, value })
        });
        if (!response.ok) throw new Error(`Find element failed: ${response.status}`);
        const data = await response.json();
        return data.value['element-6066-11e4-a52e-4f735466cecf'];
    }

    async executeScript(sessionId, script, args = []) {
        const response = await fetch(`${this.baseUrl}/session/${sessionId}/execute/sync`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ script, args })
        });
        if (!response.ok) throw new Error(`Execute script failed: ${response.status}`);
        const data = await response.json();
        return data.value;
    }

    async screenshot(sessionId) {
        const response = await fetch(`${this.baseUrl}/session/${sessionId}/screenshot`);
        if (!response.ok) throw new Error(`Screenshot failed: ${response.status}`);
        const data = await response.json();
        return data.value;
    }

    async quit(sessionId) {
        const response = await fetch(`${this.baseUrl}/session/${sessionId}`, {
            method: 'DELETE'
        });
        if (!response.ok) throw new Error(`Quit failed: ${response.status}`);
    }
}

/**
 * MCP (Model Context Protocol) Client
 */
export class MCPBridge {
    constructor(host = 'localhost', port = 9400) {
        this.baseUrl = `http://${host}:${port}/mcp`;
    }

    async info() {
        const response = await fetch(`${this.baseUrl}/info`);
        if (!response.ok) throw new Error(`Get info failed: ${response.status}`);
        return await response.json();
    }

    async listTools() {
        const response = await fetch(`${this.baseUrl}/tools`);
        if (!response.ok) throw new Error(`List tools failed: ${response.status}`);
        return await response.json();
    }

    async callTool(name, args) {
        const response = await fetch(`${this.baseUrl}/tools/call`, {
            method: 'POST',
            headers: { 'Content-Type': 'application/json' },
            body: JSON.stringify({ name, arguments: args })
        });
        if (!response.ok) throw new Error(`Call tool failed: ${response.status}`);
        return await response.json();
    }

    async listResources() {
        const response = await fetch(`${this.baseUrl}/resources`);
        if (!response.ok) throw new Error(`List resources failed: ${response.status}`);
        return await response.json();
    }

    // Convenience methods
    async browse(url) {
        const result = await this.callTool('browse', { url });
        return result.content[0].text;
    }

    async screenshot() {
        const result = await this.callTool('screenshot', {});
        return result.content[0].data;
    }

    async extract(selector) {
        const result = await this.callTool('extract', { selector });
        return result.content[0].text;
    }

    async wait(selector, timeout = 10000) {
        const result = await this.callTool('wait', { selector, timeout });
        return result.content[0].text;
    }

    async getHtml() {
        const result = await this.callTool('get_html', {});
        return result.content[0].text;
    }

    async getText() {
        const result = await this.callTool('get_text', {});
        return result.content[0].text;
    }
}

// CommonJS export support
if (typeof module !== 'undefined' && module.exports) {
    module.exports = { WebViewBridge, WebDriverBridge, MCPBridge };
}

File: sdk/python/webview_bridge.py
"""
WebView Bridge Python SDK
A simple Python client for WebView Bridge server.

Usage:
    from webview_bridge import WebViewBridge
    
    bridge = WebViewBridge()  # Connect to localhost:9400
    
    # Create session
    session_id = bridge.create_session()
    
    # Navigate
    bridge.navigate(session_id, "https://example.com")
    
    # Execute JavaScript
    result = bridge.execute(session_id, "return document.title")
    
    # Take screenshot
    screenshot_base64 = bridge.screenshot(session_id)
    
    # Close session
    bridge.close(session_id)
"""

import requests
from typing import Optional, Any, Dict, List


class WebViewBridge:
    """WebView Bridge Python SDK Client"""
    
    def __init__(self, host: str = "localhost", port: int = 9400):
        self.base_url = f"http://{host}:{port}"
        self.session = requests.Session()
    
    def health(self) -> bool:
        """Check if the server is healthy."""
        try:
            response = self.session.get(f"{self.base_url}/health")
            return response.status_code == 200
        except Exception:
            return False
    
    def create_session(self, profile: str = "default", headless: bool = False) -> str:
        """Create a new browser session."""
        response = self.session.post(
            f"{self.base_url}/create",
            json={"profile": profile, "headless": headless}
        )
        response.raise_for_status()
        return response.json()["id"]
    
    def navigate(self, session_id: str, url: str) -> bool:
        """Navigate to a URL."""
        response = self.session.post(
            f"{self.base_url}/navigate/{session_id}",
            json={"url": url}
        )
        response.raise_for_status()
        return response.json().get("success", True)
    
    def execute(self, session_id: str, script: str) -> Any:
        """Execute JavaScript in the page context."""
        response = self.session.post(
            f"{self.base_url}/execute/{session_id}",
            json={"script": script}
        )
        response.raise_for_status()
        return response.json().get("result")
    
    def get_status(self, session_id: str) -> Dict[str, Any]:
        """Get session status."""
        response = self.session.get(f"{self.base_url}/status/{session_id}")
        response.raise_for_status()
        return response.json()
    
    def snapshot(self, session_id: str, format: str = "html") -> str:
        """Get page snapshot (HTML or text)."""
        response = self.session.get(
            f"{self.base_url}/snapshot/{session_id}",
            params={"format": format}
        )
        response.raise_for_status()
        return response.json().get("content", "")
    
    def screenshot(self, session_id: str) -> str:
        """Take a screenshot (returns base64 PNG)."""
        response = self.session.get(f"{self.base_url}/screenshot/{session_id}")
        response.raise_for_status()
        return response.json().get("image", "")
    
    def get_cookies(self, session_id: str) -> List[Dict[str, Any]]:
        """Get all cookies."""
        response = self.session.get(f"{self.base_url}/cookies/{session_id}")
        response.raise_for_status()
        return response.json().get("cookies", [])
    
    def set_cookies(self, session_id: str, cookies: List[Dict[str, Any]]) -> bool:
        """Set cookies."""
        response = self.session.post(
            f"{self.base_url}/cookies/{session_id}",
            json={"cookies": cookies}
        )
        response.raise_for_status()
        return True
    
    def wait_for_selector(
        self, session_id: str, selector: str, timeout_ms: int = 10000
    ) -> bool:
        """Wait for an element to appear."""
        response = self.session.post(
            f"{self.base_url}/wait/{session_id}",
            json={"selector": selector, "timeout_ms": timeout_ms}
        )
        response.raise_for_status()
        return response.json().get("found", False)
    
    def extract(
        self, session_id: str, selector: str, attribute: str = "text"
    ) -> List[str]:
        """Extract content from elements."""
        response = self.session.post(
            f"{self.base_url}/extract/{session_id}",
            json={"selector": selector, "attribute": attribute, "all": True}
        )
        response.raise_for_status()
        return response.json().get("results", [])
    
    def close(self, session_id: str) -> bool:
        """Close a session."""
        response = self.session.delete(f"{self.base_url}/close/{session_id}")
        response.raise_for_status()
        return True


class WebDriverBridge:
    """WebDriver Protocol Compatible Client (for Selenium compatibility)"""
    
    def __init__(self, host: str = "localhost", port: int = 9400):
        self.base_url = f"http://{host}:{port}/wd/hub"
        self.session = requests.Session()
    
    def new_session(self) -> str:
        """Create a new WebDriver session."""
        response = self.session.post(
            f"{self.base_url}/session",
            json={"capabilities": {}}
        )
        response.raise_for_status()
        return response.json()["value"]["sessionId"]
    
    def navigate(self, session_id: str, url: str) -> None:
        """Navigate to URL."""
        response = self.session.post(
            f"{self.base_url}/session/{session_id}/url",
            json={"url": url}
        )
        response.raise_for_status()
    
    def get_url(self, session_id: str) -> str:
        """Get current URL."""
        response = self.session.get(f"{self.base_url}/session/{session_id}/url")
        response.raise_for_status()
        return response.json()["value"]
    
    def get_title(self, session_id: str) -> str:
        """Get page title."""
        response = self.session.get(f"{self.base_url}/session/{session_id}/title")
        response.raise_for_status()
        return response.json()["value"]
    
    def find_element(self, session_id: str, using: str, value: str) -> str:
        """Find element by selector."""
        response = self.session.post(
            f"{self.base_url}/session/{session_id}/element",
            json={"using": using, "value": value}
        )
        response.raise_for_status()
        return response.json()["value"]["element-6066-11e4-a52e-4f735466cecf"]
    
    def find_elements(self, session_id: str, using: str, value: str) -> List[str]:
        """Find elements by selector."""
        response = self.session.post(
            f"{self.base_url}/session/{session_id}/elements",
            json={"using": using, "value": value}
        )
        response.raise_for_status()
        return [
            el["element-6066-11e4-a52e-4f735466cecf"]
            for el in response.json()["value"]
        ]
    
    def execute_script(self, session_id: str, script: str, args: List = None) -> Any:
        """Execute JavaScript."""
        response = self.session.post(
            f"{self.base_url}/session/{session_id}/execute/sync",
            json={"script": script, "args": args or []}
        )
        response.raise_for_status()
        return response.json()["value"]
    
    def screenshot(self, session_id: str) -> str:
        """Take screenshot."""
        response = self.session.get(
            f"{self.base_url}/session/{session_id}/screenshot"
        )
        response.raise_for_status()
        return response.json()["value"]
    
    def quit(self, session_id: str) -> None:
        """Quit session."""
        response = self.session.delete(f"{self.base_url}/session/{session_id}")
        response.raise_for_status()


class MCPBridge:
    """MCP (Model Context Protocol) Client"""
    
    def __init__(self, host: str = "localhost", port: int = 9400):
        self.base_url = f"http://{host}:{port}/mcp"
        self.session = requests.Session()
    
    def info(self) -> Dict[str, str]:
        """Get server info."""
        response = self.session.get(f"{self.base_url}/info")
        response.raise_for_status()
        return response.json()
    
    def list_tools(self) -> List[Dict[str, Any]]:
        """List available tools."""
        response = self.session.get(f"{self.base_url}/tools")
        response.raise_for_status()
        return response.json()
    
    def call_tool(self, name: str, arguments: Dict[str, Any]) -> Dict[str, Any]:
        """Call a tool."""
        response = self.session.post(
            f"{self.base_url}/tools/call",
            json={"name": name, "arguments": arguments}
        )
        response.raise_for_status()
        return response.json()
    
    def list_resources(self) -> List[Dict[str, Any]]:
        """List available resources."""
        response = self.session.get(f"{self.base_url}/resources")
        response.raise_for_status()
        return response.json()
    
    def read_resource(self, uri: str) -> Dict[str, Any]:
        """Read a resource."""
        response = self.session.post(
            f"{self.base_url}/resources/read",
            json={"uri": uri}
        )
        response.raise_for_status()
        return response.json()
    
    # Convenience methods
    def browse(self, url: str) -> str:
        """Navigate to URL."""
        result = self.call_tool("browse", {"url": url})
        return result["content"][0]["text"]
    
    def screenshot(self) -> str:
        """Take screenshot (base64)."""
        result = self.call_tool("screenshot", {})
        return result["content"][0]["data"]
    
    def extract(self, selector: str) -> str:
        """Extract text from selector."""
        result = self.call_tool("extract", {"selector": selector})
        return result["content"][0]["text"]
    
    def wait(self, selector: str, timeout: int = 10000) -> str:
        """Wait for element."""
        result = self.call_tool("wait", {"selector": selector, "timeout": timeout})
        return result["content"][0]["text"]
    
    def get_html(self) -> str:
        """Get page HTML."""
        result = self.call_tool("get_html", {})
        return result["content"][0]["text"]
    
    def get_text(self) -> str:
        """Get page text."""
        result = self.call_tool("get_text", {})
        return result["content"][0]["text"]


# Example usage
if __name__ == "__main__":
    # Test basic connectivity
    bridge = WebViewBridge()
    
    if bridge.health():
        print("✅ WebView Bridge is running!")
        
        # Create session
        session_id = bridge.create_session()
        print(f"Session created: {session_id}")
        
        # Navigate
        bridge.navigate(session_id, "https://example.com")
        print("Navigated to example.com")
        
        # Get title
        title = bridge.execute(session_id, "return document.title")
        print(f"Title: {title}")
        
        # Close
        bridge.close(session_id)
        print("Session closed")
    else:
        print("❌ WebView Bridge is not running!")
